### PR TITLE
PhysicalFilter: add RegFieldDesc and scaladoc

### PR DESCRIPTION
**Pull request by mwachs5**
**Related issue**: 

**Type of change**:  other enhancement

**Impact**: API modification

**Development Phase**: implementation

**Release Notes**

Add RegFieldDesc and scaladoc to PhysicalFilter Tile Link Device

### CODE REVIEW

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 1]
```diff
@@ -23,53 +25,103 @@ class DevicePMP(params: DevicePMPParams) extends GenericParameterizedBundle(para
...
-    def field(bits: Int, reg: UInt, lock: Bool = l(0)) =
+  def fields(blockAddress: Bool, initial: PMPInitialValue): Seq[RegField] = {
+    val initialInts = DevicePMP.getInitialValueInts(params.addressBits, params.pageBits, Some(initial))
+
+    val lDesc = RegFieldDesc("l",
+      "Lock bit. When set, prevents modification to other fields in the register. Cannot be modified if l bit is set.",
+      reset = Some(initialInts.l),
+      wrType = Some(RegFieldWrType.MODIFY))
+
+    val aDesc = RegFieldDesc("a", "Access bit. When clear, this PMP does not filter anything. When set, Top-of-Range (TOR) filtering is applied by this PMP. Cannot be modified if l bit is set.",
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:36]:
```suggestion
    val aDesc = RegFieldDesc("a", "Address match mode. When clear, this PMP does not match any address. When set, Top-of-Range (TOR) matching is applied by this PMP. Cannot be modified if lock bit is set.",
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 2]
```diff
@@ -23,53 +25,103 @@ class DevicePMP(params: DevicePMPParams) extends GenericParameterizedBundle(para
...
+      reset = Some(initialInts.a),
+      wrType = Some(RegFieldWrType.MODIFY))
+
+    val rDesc = RegFieldDesc("r", "Read bit. When set grants read access to the matching address range. Cannot be modified if l bit is set.",
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:40]:
```suggestion
    val rDesc = RegFieldDesc("r", "Read bit. When set grants read access to the matching address range. Cannot be modified if lock bit is set.",
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 3]
```diff
@@ -23,53 +25,103 @@ class DevicePMP(params: DevicePMPParams) extends GenericParameterizedBundle(para
...
+      reset = Some(initialInts.r),
+      wrType = Some(RegFieldWrType.MODIFY))
+
+    val wDesc = RegFieldDesc("w", "Write bit. When set grants write access to the matching address range. Cannot be modified if l bit is set.",
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:44]:
```suggestion
    val wDesc = RegFieldDesc("w", "Write bit. When set grants write access to the matching address range. Cannot be modified if lock bit is set.",
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 4]
```diff
@@ -23,53 +25,103 @@ class DevicePMP(params: DevicePMPParams) extends GenericParameterizedBundle(para
...
+      reset = Some(initialInts.w),
+      wrType = Some(RegFieldWrType.MODIFY))
+
+    val addrHiDesc = RegFieldDesc("addr_hi", "Page address. Specifies top-of-range page address for this PMP and bottom-of-range address for following PMP. Cannot be modified if `l` bit is set, or if `a` bit and `l` bit is set on the subsequent PMP.",
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:48]:
```suggestion
    val addrHiDesc = RegFieldDesc("addr_hi", "Page address. Specifies top-of-range page address for this PMP and bottom-of-range address for following PMP. Cannot be modified if lock bit is set, or if address match mode is TOR and lock bit is set on the subsequent PMP.",
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 5]
```diff
@@ -23,53 +25,103 @@ class DevicePMP(params: DevicePMPParams) extends GenericParameterizedBundle(para
...
+      field(1, l, lDesc))
   }
 }
 
+
+/* Initial value of the PMP registers
+ *  
+ *  @param address - Initial value of the address field. Will be shifted down by pageBits before being stored in the PMP register addr_hi register.
+ *  @param l Initial value of the l (locked) bit. True means no modifications of other fields are allowed.
+ *  @param a Initial value of the a (address match type) bit. False means disabled, True means use TOR address matching. Also disables modification of previous PMP address.
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:77]:
```suggestion
 *  @param a Initial value of the a (address match mode) bit. False means disabled, True means use TOR address matching. Also disables modification of previous PMP address.
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 6]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
   val size = (((pmpRegisters.size * 8) + page - 1) / page) * page
 
-  require (!pmpRegisters.isEmpty)
-  require (controlAddress > 0)
-  require (controlAddress % size == 0)
-  require (controlBeatBytes > 0 && isPow2(controlBeatBytes))
+  require (!pmpRegisters.isEmpty,
+    "Must specify at least one Device PMP register")
+  require (controlAddress > 0,
+    "Must specify a non-zero control address for PhysicalFilter")
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:138]:
```suggestion
    s"Must specify a positive, non-zero control address for PhysicalFilter, not ${controlAddress}")
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 7]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
+  require (controlAddress % size == 0,
+    s"PhysicalFilter Control address must be aligned to its size, not size ${size} at ${controlAddress}")
+  require (controlBeatBytes > 0 && isPow2(controlBeatBytes),
+    s"PhyiscalFilter control beat bytes must be a positive power of two, not ${controlBeatBytes}")
 }
 
+/** The PhyisicalFilter provides physical memory protection for Tile Link bus traffic, granting or denying accesses based on address and access type.
+  *  
+  *  The PhysicalFilter uses a set of DevicePMP registers to control whether
+  *  accesses of certain types are allowed to proceed or denied.
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:148]:
```suggestion
  *  accesses of certain types are allowed to proceed or will be denied.
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 8]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
+  *  For transactions in flight, the PhysicalFilter will only prevent acquisition of NEW permissions;
+  *  it will not shoot down permissions acquired previously.
+  *  
+  *  The blocking behavior is controlled by a series of PMP registers which are accessible via memory mapped reads and writes.
+  *  The list is a priority allow list. If no PMP matches the transaction will be denied. Otherwise the first PMP which is active
+  *  and address matches is compared against the requested read and/or write permissions. 
+  *  When an access is denied they Phyiscal Filter crafts and responds with a Tile Link Denied response message.
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:155]:
```suggestion
  *  When an access is denied the Phyiscal Filter crafts and responds with a Tile Link Denied response message.
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 9]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
+  *  
+  *  When a device PMP register's a bit is set, it is enabled and Top of Range (TOR) matching is applied.
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:157]:
```suggestion
  *  When a device PMP register's `a` bit is set, it is enabled and Top of Range (TOR) matching is applied.
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 10]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
+  *  For a given PMP register, the associated address register forms the top of the
+  *  address range, and the preceding PMP address register forms the bottom of the
+  *  address range.  If PMP[i]'s `a` field is set to TOR, the entry matches
+  *  any address `y` such that PMP[i-1].address <= y < PMP[i].address
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:161]:
```suggestion
  *  any address `y` such that `PMP[i-1].address <= y < PMP[i].address` .
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 11]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
+  *  If PMP[0].a is set (TOR is applied), zero is used for the lower bound, and so
+  *  it matches any address `y` < PMP[0].address.
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:163]:
```suggestion
  *  it matches any address `y < PMP[0].address`.
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 12]
```diff
@@ -80,12 +132,49 @@ case class PhysicalFilterParams(
...
+  *  Note that the addresses bits found in the PMP registers are addr_hi, or page address.
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:164]:
```suggestion
  *  Note that the addresses bits stored in the PMP registers are addr_hi, or page address.
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 13]
```diff
@@ -122,9 +213,11 @@ class PhysicalFilter(params: PhysicalFilterParams)(implicit p: Parameters) exten
       val needR = in.a.bits.opcode =/= TLMessages.PutFullData &&
                   in.a.bits.opcode =/= TLMessages.PutPartialData
       val lt = Bool(false) +: pmps.map(in.a.bits.address < _.address)
+      // sel[i] is true if PMP[i].a is set and the PMP[i-1].address <= address < PMP[i].address
```
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:216]:
```suggestion
      // sel[i] is true if PMP[i].a is set and PMP[i-1].address <= address < PMP[i].address
```

File: src/main/scala/devices/tilelink/PhysicalFilter.scala [hunk 14]
```diff
@@ -23,53 +25,103 @@ class DevicePMP(params: DevicePMPParams) extends GenericParameterizedBundle(para
...
     out
   }
-}
 
-/** PhysicalFilter uses a set of DevicePMP registers to control whether
-  * accesses of certain types are allowed to proceed or denied. The Filter
-  * will only prevent acquisition of NEW permissions; it will not shoot
-  * down permissions Acquired previously.
-  */
+  /** Helper to convert from config booleans to integer values needed for the DevicePMP Bundles */
```
Comment (hcook) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:109]: > from config booleans

but takes `Int`s and an `Option[PMPInitialValue]` as an argument?

Comment (mwachs5) [reply src/main/scala/devices/tilelink/PhysicalFilter.scala:109]: inside the PMPInitial value are Booleans and I want Ints.... will clarify

**Review by hcook — APPROVED** [review]
I don't know how the block actually works but the added messages and comments all look self-consistent.

File: src/main/scala/devices/tilelink/PhysicalFilter.scala (hunk 14 above)
Comment (mwachs5) [inline src/main/scala/devices/tilelink/PhysicalFilter.scala:109]:
```suggestion
  /** Helper to convert from Booleans in the optional PMPInitialValue to Int/BigInt values needed to create the UInts in a DevicePMP Bundle */

**Pull request merged**